core/base/src/TColor.cxx

```diff
@@ -1980,6 +1980,27 @@ Int_t TColor::GetColor(Int_t r, Int_t g, Int_t b, Float_t a)
    return color->GetNumber();
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Static method returning color number for a color specified by name.
+///
+/// If specified color name does not exist -1 is returned.
+
+Int_t TColor::GetColorByName(const char *colorname)
+{
+   TObjArray *colors = (TObjArray*) gROOT->GetListOfColors();
+   Int_t ncolors = colors->GetSize();
+   TColor *color = 0;
+   TString colname;
+
+   for (Int_t i = 0; i<ncolors; i++) {
+      color = (TColor*)colors->At(i);
+      colname = color->GetName();
+      if (!colname.Index(colorname)) return i;
+   }
+
+   return -1;
```

Comment:
```suggestion
const auto colors = TRangeDynCast<TColor>(gROOT->GetListOfColors());
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorName; }); it != colors.end()){
    return static_cast<Int_t>(std::distance(colors.begin(), it));
} else {
    return -1;
}
```

Reply:
hmm.... is it faster ? I do not think it is more clear...

Reply:
imho it is more clear. Plus, it doesn't use c-style casts and doesn't need to use TString.

Also, returning `-1` is a bit ugly, throwing an exception would be better. But I understand that is done to keep compatibility with the rest of the interfaces of this class.

Reply:
your code gives me:
```
[ 73%] Building CXX object core/CMakeFiles/Core.dir/base/src/TColor.cxx.o
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:105: error: variable 'colorname' cannot be implicitly captured in a lambda with no capture-default specified
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                                                                        ^
/Users/couet/git/couet-root/core/base/src/TColor.cxx:1988:42: note: 'colorname' declared here
Int_t TColor::GetColorByName(const char *colorname)
                                         ^
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:58: note: lambda expression begins here
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                         ^
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:59: note: capture 'colorname' by value
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                          ^
                                                          colorname
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:59: note: capture 'colorname' by reference
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                          ^
                                                          &colorname
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:59: note: default capture by value
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                          ^
                                                          =
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2005:59: note: default capture by reference
if (auto it = std::find_if(colors.begin(), colors.end(), [](TColor *color) { return color->GetName() == colorname; }); it != colors.end()){
                                                          ^
                                                          &
/Users/couet/git/couet-root/core/base/src/TColor.cxx:2006:31: error: no matching function for call to 'distance'
    return static_cast<Int_t>(std::distance(colors.begin(), it));
                              ^~~~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__iterator/distance.h:51:1: note: candidate template ignored: substitution failure [with _InputIter = ROOT::Internal::TypedIter<TColor *, TIter, true>]: no type named 'difference_type' in 'std::iterator_traits<ROOT::Internal::TypedIter<TColor *, TIter, true>>'
distance(_InputIter __first, _InputIter __last)
^
2 errors generated.

```

Reply:
Yep, there's just a missing capture in the lambda as the error suggests.
This will do
```
[&colorName](TColor *) { ... }
```

Also, for my code to work `colorName` should be an `std::string` for the `operator==`. If you want to keep the `const char *` interface then the comparison should be done via `strcmp`

Reply:
Therefore, at the end of the day, I am not sure we gain clarity... but that's a matter of taste of course.